### ISSUE TITLE
Optimisation: `activate <environment>` takes 1.9 seconds

### ISSUE THREAD

**Issue by HaleTom** [conda/conda#8609]
I've setup an environment for neural network deep learning, which contains tensorflow, numpy and pytorch (among others).

Activating this environment via `conda activate nndl` is *slooow*, taking about 2 seconds on a quite idle Intel Core i5 6500U, ([Asus UX330 specs](http://www.zenbook.org/ux330/)).

Given that all this command does is set some environment variables, this is orders of magnitude slower than I'd expect.

I avoid using something like tools like `autocd` to automatically set environments as I move in and out of directories because the delay is too noticeable.

The annoyance is exacerbated by environment variables not being exported to subprocesses. As a work-around, I would like to `activate` once, then run `tmux`  and have all windows also activated.  If there is a solution for exporting the activated variables, please let me know.

Note: #7239 is about the speed of the solver, this is about the speed of the `source`d script which sets up environment variables.

**Comment by mbargull** [comment]
> I've setup an environment for neural network deep learning, which contains tensorflow, numpy and pytorch (among others).

Without any of the information that is asked for in the issue templates, it is not possible to determinate what is amiss here.

> *slooow*, taking about 2 seconds

Yes, that's really slow. We do have somewhat hard limits on how fast the environment activation can get since launching CPython and executing a bit of Python code just takes it time. But that limit should at most be a couple of 100 milliseconds.
What you are seeing is *likely* ( => can't really tell due to insufficient information) due to
- some package's de-/activation scripts being executed and delaying things (check the `etc/conda/{de,}activate.d` directories of the deactivated/activated environments),
- slow I/O due to environment being on a network mount or the like,
- something else externally interfering.

> The annoyance is exacerbated by environment variables not being exported to subprocesses.

This statement is not correct. Environment variables are being exported and thus are available to subprocesses (this excludes `conda activate`/`conda deactivate` since the `conda` there is a function for the shell itself, hence not `export`able).
What you are seeing is just due to how `tmux` works: Subsequent launches of `tmux` inherit from the first `tmux` instance that you started. If you want to test this, just close all `tmux` instances, run `conda activate && tmux new` and inspect the environment variables in there and subsequently launched `tmux` sessions.
I'm not sure what the best strategy for you might be, but you could in `tmux` checkout the `setenv` and/or `default-shell` options (e.g., with launching `bash` with a custom `init-file`). If you come up with a solution that works for you, feel free to share it with other users here.
Handling this in `conda` itself is either not possible or out of scope of this project, IMO.

**Comment by HaleTom** [comment]
@mbargull Thanks for clarifying about `tmux`. 

I can't generate the `conda_debug_output.txt`:
```
% CONDA_INSTRUMENTATION_ENABLED=1 $XDG_DATA_HOME/miniconda3/bin/activate -vv nndl
activate does not accept more than one argument:
['-vv', 'nndl']
```

**How do I generate the output the template asks for?**

Other template requested information is in the next comment. (Sorry, I used the wrong template initially)
-----------

`activate` has two commands:

1. `\. "$_CONDA_ROOT/etc/profile.d/conda.sh" || return $?`
1. `conda activate "$@"`

I present timings for both below.


## \. "$_CONDA_ROOT/etc/profile.d/conda.sh"
Here are some timings with both `{ba,z}sh`:

```
% bash --noprofile --norc -c 'start=$(date +"%s.%N"); \source $XDG_DATA_HOME/miniconda3/etc/profile.d/conda.sh; end=$(date +"%s.%N"); echo "$end-$start" | bc'
.003859784
```

This is high, but is due to the testing method itself, here's a null test:

```
% bash --noprofile --norc -c 'start=$(date +"%s.%N");  end=$(date +"%s.%N"); echo "$end-$start" | bc' 
.002608824
```
`zsh` is more accurate:
```
% zsh -fc 'typeset -F SECONDS=0; \source $XDG_DATA_HOME/miniconda3/etc/profile.d/conda.sh; echo $SECONDS'
0.0009380000
```

From the above, `conda.sh`'s runtime is acceptable.

## `conda activate "$@"`

Timing is more precise in `zsh`, so I show only that:

```
% zsh -fc '\source $XDG_DATA_HOME/miniconda3/etc/profile.d/conda.sh; typeset -F SECONDS=0; conda activate nndl; echo $SECONDS'
1.6786250000
% zsh -fc '\source $XDG_DATA_HOME/miniconda3/etc/profile.d/conda.sh; typeset -F SECONDS=0; conda activate nndl; echo $SECONDS'
1.6548370000
% zsh -fc '\source $XDG_DATA_HOME/miniconda3/etc/profile.d/conda.sh; typeset -F SECONDS=0; conda activate nndl; echo $SECONDS'
1.6106840000
% uptime                         
 12:34:46 up 6 days, 15:55,  2 users,  load average: 0.81, 0.76, 0.76
% 
```

Note: I have hidden the `{,de}activate` directories:
```
% ls -l "$XDG_DATA_HOME/miniconda3/etc/conda"
total 0
drwxr-xr-x 1 ravi ravi 46 Oct 17  2018 activate.d.hidden/
drwxr-xr-x 1 ravi ravi 46 Oct 17  2018 deactivate.d.hidden/
```
--- 

**I believe 1.65 seconds is unacceptable for something which simply sets some environment variables.**

**Comment by HaleTom** [comment]
```
% conda info

     active environment : None
       user config file : /home/ravi/.condarc
 populated config files : /home/ravi/.condarc
          conda version : 4.6.14
    conda-build version : not installed
         python version : 3.7.0.final.0
       base environment : /home/ravi/.local/share/miniconda3  (writable)
           channel URLs : https://conda.anaconda.org/fastai/linux-64
                          https://conda.anaconda.org/fastai/noarch
                          https://conda.anaconda.org/pytorch/linux-64
                          https://conda.anaconda.org/pytorch/noarch
                          https://conda.anaconda.org/intel/linux-64
                          https://conda.anaconda.org/intel/noarch
                          https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/free/linux-64
                          https://repo.anaconda.com/pkgs/free/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/ravi/.local/share/miniconda3/pkgs
                          /home/ravi/.conda/pkgs
       envs directories : /home/ravi/.local/share/miniconda3/envs
                          /home/ravi/.conda/envs
               platform : linux-64
             user-agent : conda/4.6.14 requests/2.21.0 CPython/3.7.0 Linux/4.19.34-1-MANJARO manjaro/18.0.4 glibc/2.28
                UID:GID : 1000:1000
             netrc file : /home/ravi/.netrc
           offline mode : False
```

```
% conda config --show-sources
==> /home/ravi/.condarc <==
changeps1: False
pinned_packages:
  - conda-forge::pip
  - conda-forge::python
channels:
  - fastai
  - pytorch
  - intel
  - conda-forge
  - defaults
report_errors: True
```

```
% conda list --show-channel-urls
# packages in environment at /home/ravi/.local/share/miniconda3:
#
# Name                    Version                   Build  Channel
asn1crypto                0.24.0                py37_1003    conda-forge
attrs                     19.1.0                     py_0    conda-forge
backcall                  0.1.0                      py_0    conda-forge
bleach                    3.1.0                      py_0    conda-forge
bzip2                     1.0.6                        17    intel
ca-certificates           2019.3.9             hecc5488_0    conda-forge
certifi                   2019.3.9                 py37_0    conda-forge
cffi                      1.12.3           py37h8022711_0    conda-forge
chardet                   3.0.4                 py37_1003    conda-forge
colorama                  0.4.1                      py_0    conda-forge
conda                     4.6.14                   py37_0    conda-forge
conda-env                 2.6.0                         1    conda-forge
cryptography              2.5              py37hb7f436b_1    conda-forge
decorator                 4.4.0                      py_0    conda-forge
defusedxml                0.5.0                      py_1    conda-forge
entrypoints               0.3                   py37_1000    conda-forge
gitdb2                    2.0.5                      py_0    conda-forge
gitpython                 2.1.11                     py_0    conda-forge
gmp                       6.1.2             hf484d3e_1000    conda-forge
greenlet                  0.4.15                   pypi_0    pypi
idna                      2.8                   py37_1000    conda-forge
intelpython               2019.3                        0    intel
ipykernel                 5.1.0           py37h24bf2e0_1002    conda-forge
ipython                   7.5.0            py37h24bf2e0_0    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
jedi                      0.13.3                   py37_0    conda-forge
jinja2                    2.10.1                     py_0    conda-forge
jsonschema                3.0.1                    py37_0    conda-forge
jupyter_client            5.2.4                      py_3    conda-forge
jupyter_core              4.4.0                      py_0    conda-forge
libedit                   3.1.20170329      hf8c457e_1001    conda-forge
libffi                    3.2.1                        11    intel
libgcc-ng                 7.3.0                hdf63c60_0    conda-forge
libsodium                 1.0.16                        3    intel
libstdcxx-ng              7.3.0                hdf63c60_0    conda-forge
markupsafe                1.1.1            py37h14c3975_0    conda-forge
mistune                   0.8.4           py37h14c3975_1000    conda-forge
msgpack                   0.5.6                    pypi_0    pypi
nbconvert                 5.5.0                      py_0    conda-forge
nbdime                    1.0.5                    py37_0    conda-forge
nbformat                  4.4.0                      py_1    conda-forge
ncurses                   6.1               hf484d3e_1002    conda-forge
neovim                    0.3.1                    pypi_0    pypi
notebook                  5.7.8                    py37_0    conda-forge
openssl                   1.0.2r                        0    intel
pandoc                    2.7.2                         0    conda-forge
pandocfilters             1.4.2                      py_1    conda-forge
parso                     0.4.0                      py_0    conda-forge
pexpect                   4.7.0                    py37_0    conda-forge
pickleshare               0.7.5                 py37_1000    conda-forge
pip                       18.1                     pypi_0    pypi
prometheus_client         0.6.0                      py_0    conda-forge
prompt_toolkit            2.0.9                      py_0    conda-forge
ptyprocess                0.6.0                   py_1001    conda-forge
pycosat                   0.6.3           py37h14c3975_1001    conda-forge
pycparser                 2.19                     py37_1    conda-forge
pygments                  2.3.1                      py_0    conda-forge
pynvim                    0.3.2                    pypi_0    pypi
pyopenssl                 19.0.0                   py37_0    conda-forge
pyrsistent                0.15.0           py37h516909a_0    conda-forge
pysocks                   1.6.8                 py37_1002    conda-forge
python                    3.7.0                h5001a0f_0    conda-forge
python-dateutil           2.8.0                      py_0    conda-forge
pyzmq                     18.0.1           py37hc4ba49a_1    conda-forge
readline                  7.0               hf8c457e_1001    conda-forge
requests                  2.21.0                py37_1000    conda-forge
ruamel_yaml               0.15.71         py37h14c3975_1000    conda-forge
send2trash                1.5.0                      py_0    conda-forge
setuptools                41.0.1                   py37_0    conda-forge
six                       1.12.0                py37_1000    conda-forge
smmap2                    2.0.5                      py_0    conda-forge
sqlite                    3.27.2                        2    intel
tcl                       8.6.4                        22    intel
terminado                 0.8.2                    py37_0    conda-forge
testpath                  0.4.2                   py_1001    conda-forge
tk                        8.6.9             h84994c4_1001    conda-forge
tornado                   6.0.2            py37h516909a_0    conda-forge
traitlets                 4.3.2                 py37_1000    conda-forge
urllib3                   1.24.2                   py37_0    conda-forge
wcwidth                   0.1.7                      py_1    conda-forge
webencodings              0.5.1                      py_1    conda-forge
wheel                     0.33.1                   py37_0    conda-forge
xz                        5.2.3                         2    intel
yaml                      0.1.7                         2    intel
zeromq                    4.3.1             hf484d3e_1000    conda-forge
zlib                      1.2.11                        5    intel
%
```

**Comment by HaleTom** [comment]
Packages for the `nndl` environment:

```
% conda list --show-channel-urls
# packages in environment at /home/ravi/.local/share/miniconda3/envs/nndl:
#
# Name                    Version                   Build  Channel
_tflow_1100_select        0.0.3                       mkl    defaults
absl-py                   0.7.1                    py36_0    conda-forge
asn1crypto                0.24.0             py36_intel_1  [intel]  intel
astor                     0.7.1                      py_0    conda-forge
backcall                  0.1.0              py36_intel_0  [intel]  intel
backports                 1.0                py36_intel_6  [intel]  intel
blas                      2.7                    openblas    conda-forge
bleach                    2.1.3              py36_intel_0  [intel]  intel
bottleneck                1.2.1           py36h3010b51_1001    conda-forge
bzip2                     1.0.6                  intel_14  [intel]  intel
c-ares                    1.15.0            h14c3975_1001    conda-forge
ca-certificates           2019.3.9             hecc5488_0    conda-forge
certifi                   2018.1.18          py36_intel_0  [intel]  intel
cffi                      1.11.5             py36_intel_1  [intel]  intel
chardet                   3.0.4              py36_intel_0  [intel]  intel
cryptography              2.2.2              py36_intel_1  [intel]  intel
cudatoolkit               10.0.130                      0    defaults
cudnn                     7.3.1                cuda10.0_0    defaults
cupti                     10.0.130                      0    defaults
cycler                    0.10.0             py36_intel_5  [intel]  intel
cymem                     1.31.2                   py36_0    conda-forge
cytoolz                   0.9.0.1         py36h14c3975_1001    conda-forge
dataclasses               0.6                        py_0    fastai
decorator                 4.3.0              py36_intel_0  [intel]  intel
dill                      0.2.9                    py36_0    conda-forge
entrypoints               0.2.3              py36_intel_0  [intel]  intel
fastai                    1.0.27                     py_1    fastai
fastprogress              0.1.21                     py_0    fastai
freetype                  2.9                     intel_1  [intel]  intel
future                    0.17.1                   pypi_0    pypi
gast                      0.2.2                      py_0    conda-forge
get_terminal_size         1.0.0              py36_intel_5  [intel]  intel
grpcio                    1.14.1           py36hd60e7a3_0    conda-forge
h5py                      2.7.1           np114py36_intel_0  [intel]  intel
hdf5                      1.10.1                  intel_0  [intel]  intel
html5lib                  1.0.1              py36_intel_1  [intel]  intel
icc_rt                    2019.3                intel_199    intel
idna                      2.6                py36_intel_0  [intel]  intel
intel-openmp              2019.3                intel_199    intel
intelpython               2018.0.3                      0    intel
ipykernel                 4.6.1              py36_intel_0  [intel]  intel
ipython                   6.3.1              py36_intel_1  [intel]  intel
ipython_genutils          0.2.0              py36_intel_0  [intel]  intel
ipywidgets                7.0.0              py36_intel_0  [intel]  intel
jedi                      0.12.0             py36_intel_0  [intel]  intel
jinja2                    2.9.6              py36_intel_0  [intel]  intel
jsonschema                2.6.0              py36_intel_0  [intel]  intel
jupyter                   1.0.0              py36_intel_5  [intel]  intel
jupyter_client            5.1.0              py36_intel_0  [intel]  intel
jupyter_console           5.1.0              py36_intel_0  [intel]  intel
jupyter_core              4.4.0              py36_intel_1  [intel]  intel
keras                     2.2.2                    py36_0    conda-forge
keras-applications        1.0.4                      py_1    conda-forge
keras-base                2.2.2                    py36_0    defaults
keras-preprocessing       1.0.2                      py_1    conda-forge
kiwisolver                1.0.1                    py36_1    intel
libblas                   3.8.0                7_openblas    conda-forge
libcblas                  3.8.0                7_openblas    conda-forge
libedit                   3.1.20170329                  0    conda-forge
libffi                    3.2.1                   intel_8  [intel]  intel
libgcc-ng                 7.3.0                hdf63c60_0    conda-forge
libgfortran               3.0.0                         1    conda-forge
libgfortran-ng            7.2.0                hdf63c60_3    conda-forge
libgpuarray               0.7.6             h14c3975_1003    conda-forge
libjpeg-turbo             2.0.2                h14c3975_0    conda-forge
liblapack                 3.8.0                7_openblas    conda-forge
liblapacke                3.8.0                7_openblas    conda-forge
libpng                    1.6.34                  intel_1  [intel]  intel
libprotobuf               3.7.1                h8b12597_0    conda-forge
libsodium                 1.0.16                  intel_1  [intel]  intel
libstdcxx-ng              7.3.0                hdf63c60_0    conda-forge
mako                      1.0.7                      py_1    conda-forge
markdown                  2.6.11                     py_0    conda-forge
markupsafe                1.0                py36_intel_0  [intel]  intel
matplotlib                2.2.2           np114py36_intel_1  [intel]  intel
mistune                   0.7.4              py36_intel_1  [intel]  intel
mkl                       2019.3                intel_199    intel
mkl_fft                   1.0.2           np114py36_intel_0  [intel]  intel
mkl_random                1.0.1           np114py36_intel_0  [intel]  intel
msgpack-numpy             0.4.4.2                    py_1    conda-forge
msgpack-python            0.6.1            py36h6bb024c_0    conda-forge
murmurhash                1.0.0            py36hfc679d8_0    conda-forge
nbconvert                 5.2.1              py36_intel_0  [intel]  intel
nbformat                  4.4.0              py36_intel_0  [intel]  intel
ncurses                   5.9                          10    conda-forge
ninja                     1.9.0                h6bb024c_0    conda-forge
notebook                  5.2.2              py36_intel_0  [intel]  intel
numexpr                   2.6.4           np114py36_intel_0  [intel]  intel
numpy                     1.14.3             py36_intel_0  [intel]  intel
numpy-base                1.15.0           py36h1793315_0    defaults
olefile                   0.44               py36_intel_0  [intel]  intel
openblas                  0.3.5             h9ac9557_1001    conda-forge
openmp                    2018.0.3                intel_0    intel
openssl                   1.0.2o                  intel_0  [intel]  intel
pandas                    0.22.0          np114py36_intel_4  [intel]  intel
pandocfilters             1.4.1              py36_intel_0  [intel]  intel
parso                     0.2.0              py36_intel_0  [intel]  intel
path.py                   11.0.1             py36_intel_0  [intel]  intel
pexpect                   4.2.1              py36_intel_1  [intel]  intel
pickleshare               0.7.4              py36_intel_1  [intel]  intel
pillow                    5.3.0            py36hc736899_0    conda-forge
pillow-simd               5.3.0.post0              pypi_0    pypi
pip                       19.1                     py36_0    conda-forge
plac                      1.0.0                      py_0    conda-forge
portpicker                1.1.1                    py36_0    intel
preshed                   1.0.1           py36hf484d3e_1000    conda-forge
prompt_toolkit            1.0.15             py36_intel_0  [intel]  intel
protobuf                  3.5.2              py36_intel_0  [intel]  intel
psutil                    5.4.8                    pypi_0    pypi
ptyprocess                0.5.2              py36_intel_0  [intel]  intel
pycparser                 2.18               py36_intel_0  [intel]  intel
pygments                  2.2.0              py36_intel_1  [intel]  intel
pygpu                     0.7.6           py36h3010b51_1000    conda-forge
pyopenssl                 17.5.0             py36_intel_0  [intel]  intel
pyparsing                 2.2.0              py36_intel_0  [intel]  intel
pysocks                   1.6.7              py36_intel_0  [intel]  intel
python                    3.6.5                         1    conda-forge
python-dateutil           2.6.0              py36_intel_3  [intel]  intel
pytorch-cpu               1.0.1               py3.6_cpu_2    pytorch
pytorch-nightly-cpu       1.1.0.dev20190425     py3.6_cpu_0    pytorch
pytz                      2018.4             py36_intel_0  [intel]  intel
pyyaml                    3.12               py36_intel_3  [intel]  intel
pyzmq                     16.0.2             py36_intel_4  [intel]  intel
readline                  7.0                           0    conda-forge
regex                     2018.01.10      py36h14c3975_1000    fastai
requests                  2.18.4             py36_intel_0  [intel]  intel
requests-toolbelt         0.9.1                    pypi_0    pypi
scipy                     1.0.1           np114py36_intel_3  [intel]  intel
setuptools                39.0.1                   py36_0    intel
simplegeneric             0.8.1              py36_intel_5  [intel]  intel
six                       1.11.0                   py36_2    intel
spacy                     2.0.16           py36h962f231_0    defaults
sqlite                    3.20.1                  intel_0  [intel]  intel
tbb                       2019.4                intel_199    intel
tbb4py                    2019.4             py36_intel_0    intel
tcl                       8.6.4                  intel_19  [intel]  intel
tensorboard               1.13.1                   py36_0    conda-forge
tensorflow                1.3.1               np114py36_5    intel
termcolor                 1.1.0                      py_2    conda-forge
terminado                 0.8.1              py36_intel_0  [intel]  intel
testpath                  0.3.1              py36_intel_0  [intel]  intel
theano                    1.1                py36_intel_4  [intel]  intel
thinc                     6.12.0           py36h4989274_0    defaults
tk                        8.6.4                  intel_26  [intel]  intel
toolchain                 2.3.0                         0    conda-forge
toolchain_c_linux-64      2.3.0                         0    conda-forge
toolchain_cxx_linux-64    2.3.0                         0    conda-forge
toolz                     0.9.0                      py_1    conda-forge
torchvision-cpu           0.2.1                    py36_1    pytorch
torchvision-nightly-cpu   0.2.1              pyh19dea27_0    fastai
tornado                   4.5.2              py36_intel_0  [intel]  intel
tqdm                      4.31.1                     py_0    conda-forge
traitlets                 4.3.2              py36_intel_1  [intel]  intel
typing                    3.6.4                    py36_0    conda-forge
udacity-pa                0.2.8                    pypi_0    pypi
ujson                     1.35            py36h14c3975_1001    conda-forge
urllib3                   1.22               py36_intel_0  [intel]  intel
wcwidth                   0.1.7              py36_intel_5  [intel]  intel
weakref                   1.0.post1                py36_0    intel
webencodings              0.5.1                    py36_0    intel
werkzeug                  0.14.1                   py36_0    intel
wheel                     0.31.0             py36_intel_0  [intel]  intel
widgetsnbextension        3.2.0                    py36_0    intel
wrapt                     1.11.1           py36h14c3975_0    conda-forge
xz                        5.2.3                   intel_0  [intel]  intel
yaml                      0.1.7                   intel_0  [intel]  intel
zeromq                    4.2.3                   intel_0  [intel]  intel
zlib                      1.2.11                  intel_3  [intel]  intel
```

**Comment by mbargull** [comment]
Thanks for the further information.
So, the problem is most likely the activation script from `keras`, i.e., not much we can do from the `conda` side of things. You may want to open an issue over at https://github.com/conda-forge/keras-feedstock/issues, possibly even at https://github.com/keras-team/keras/issues as the `python -c 'import keras'` command is the slow part.
If you want to confirm this, you can `set -x` (or launch `zsh`/`bash`/`sh` with `-x`) and try to observe where `conda activate` stalls. On my machine the `python -c 'import keras'` part takes about a second. Overall, the runtime is also different depending on what storage driver etc. I run it. The following shows you the `tail` of the output when running on the host itself, with `docker` and `overlay2`, and with `podman` and `fuse-overlayfs`:
<details open><summary>

```
{ cat <<'EOF'
. '/home/mbargull/code/conda/conda-git-4.6.x/etc/profile.d/conda.sh'
conda_base="$(conda info --base)"
env_name='keras-2.2.2-py36_0'
conda create -yqn "${env_name}" -c conda-forge keras==2.2.2=py36_0
bash --norc --noprofile -xec ". '${conda_base}/etc/profile.d/conda.sh' && conda activate '${env_name}'" 2>&1 | {
    "${conda_base}/bin/python" -c '
from itertools import chain
from sys import stdin, stdout
from time import time
t1 = t0 = time()
for line in chain(["=== START ===\n"], stdin, ["===  END  ===\n"]):
    t2 = time()
    stdout.write("{:8.3f} | {:8.3f} |   {}".format(t2 - t0, t2 - t1, line))
    t1 = t2
'
}
exit
EOF
} | sh -e | tail
```
</summary>

```
   0.070 |    0.000 |   ++ export CONDA_PYTHON_EXE=/home/mbargull/code/conda/conda-git-4.6.x/bin/python
   0.070 |    0.000 |   ++ CONDA_PYTHON_EXE=/home/mbargull/code/conda/conda-git-4.6.x/bin/python
   0.070 |    0.000 |   ++ . /home/mbargull/code/conda/conda-git-4.6.x/envs/keras-2.2.2-py36_0/etc/conda/activate.d/keras_activate.sh
   0.071 |    0.000 |   ++++ python /home/mbargull/code/conda/conda-git-4.6.x/envs/keras-2.2.2-py36_0/etc/keras/load_config.py
   0.092 |    0.022 |   +++ KERAS_BACKEND=tensorflow
   0.092 |    0.000 |   +++ python -c 'import keras'
   0.828 |    0.736 |   + _conda_hashr
   0.828 |    0.000 |   + case "$_CONDA_SHELL_FLAVOR" in
   0.828 |    0.000 |   + hash -r
   0.828 |    0.000 |   ===  END  ===
```
</details>
<details open><summary>

```
{ cat <<'EOF'
conda install -qy conda=4.6.14
conda_base="$(conda info --base)"
env_name='keras-2.2.2-py36_0'
conda create -yqn "${env_name}" -c conda-forge keras==2.2.2=py36_0
bash --norc --noprofile -xec ". '${conda_base}/etc/profile.d/conda.sh' && conda activate '${env_name}'" 2>&1 | {
    "${conda_base}/bin/python" -c '
from itertools import chain
from sys import stdin, stdout
from time import time
t1 = t0 = time()
for line in chain(["=== START ===\n"], stdin, ["===  END  ===\n"]):
    t2 = time()
    stdout.write("{:8.3f} | {:8.3f} |   {}".format(t2 - t0, t2 - t1, line))
    t1 = t2
'
}
exit
EOF
} | docker run --rm -i continuumio/miniconda3 sh -e | tail
```
</summary>

```
   0.067 |    0.000 |   ++ CONDA_PYTHON_EXE=/opt/conda/bin/python
   0.067 |    0.000 |   ++ . /opt/conda/envs/keras-2.2.2-py36_0/etc/conda/activate.d/keras_activate.sh
   0.067 |    0.000 |   ++++ python /opt/conda/envs/keras-2.2.2-py36_0/etc/keras/load_config.py
   0.091 |    0.024 |   +++ KERAS_BACKEND=tensorflow
   0.091 |    0.000 |   +++ python -c 'import keras'
   1.024 |    0.933 |   + __conda_hashr
   1.024 |    0.000 |   + '[' -n '' ']'
   1.024 |    0.000 |   + '[' -n '' ']'
   1.024 |    0.000 |   + hash -r
   1.024 |    0.000 |   ===  END  ===
```
</details>
<details open><summary>

```
 \
{ cat <<'EOF'
conda install -qy conda=4.6.14
conda_base="$(conda info --base)"
env_name='keras-2.2.2-py36_0'
conda create -yqn "${env_name}" -c conda-forge keras==2.2.2=py36_0
bash --norc --noprofile -xec ". '${conda_base}/etc/profile.d/conda.sh' && conda activate '${env_name}'" 2>&1 | {
    "${conda_base}/bin/python" -c '
from itertools import chain
from sys import stdin, stdout
from time import time
t1 = t0 = time()
for line in chain(["=== START ===\n"], stdin, ["===  END  ===\n"]):
    t2 = time()
    stdout.write("{:8.3f} | {:8.3f} |   {}".format(t2 - t0, t2 - t1, line))
    t1 = t2
'
}
exit
EOF
} | podman run --rm -i continuumio/miniconda3 sh -e | tail
```
</summary>

```
   0.108 |    0.000 |   ++ CONDA_PYTHON_EXE=/opt/conda/bin/python
   0.108 |    0.000 |   ++ . /opt/conda/envs/keras-2.2.2-py36_0/etc/conda/activate.d/keras_activate.sh
   0.108 |    0.000 |   ++++ python /opt/conda/envs/keras-2.2.2-py36_0/etc/keras/load_config.py
   0.147 |    0.039 |   +++ KERAS_BACKEND=tensorflow
   0.147 |    0.000 |   +++ python -c 'import keras'
   1.298 |    1.151 |   + __conda_hashr
   1.298 |    0.000 |   + '[' -n '' ']'
   1.298 |    0.000 |   + '[' -n '' ']'
   1.298 |    0.000 |   + hash -r
   1.298 |    0.000 |   ===  END  ===
```
</details>

**Comment by ijstokes** [comment]
My `conda activate foo` takes 90 seconds to complete.  This is on an ReFS striped 1TB SSD on a Windows Server EC2 r5d.8xlarge instance.  I'm going to report it as a separate issue, but thought I'd add the data point here.

```
user@host [/e/ian/repo/config] (master)
$ time conda activate foo

real    1m32.700s
user    0m0.015s
sys     0m0.061s
```

**Comment by spikerobinsonDLG** [comment]
It is much, much worse in Anaconda Navigator. It is not enough even to brew a cup of tea while waiting for Navigator to switch environments. I actually have to go and do a whole separate task while I'm waiting. The delay is in the range of 15 min - 30 min.

**Comment by Doggie52** [comment]
This should most definitely not be closed. I'm facing the same types of startup times ~2s on my Windows 11 running conda 4.12.0.

**Comment by HaleTom** [comment]
> My `conda activate foo` takes 90 seconds to complete.  This is on an ReFS striped 1TB SSD on a Windows Server EC2 r5d.8xlarge instance.  I'm going to report it as a separate issue, but thought I'd add the data point here.

What is the issue number that you raised?

**Comment by liuliping0315** [comment]
Sorry to reply the closed post.
I have the same issue on a machine, and the tensorflow/keras version is 2.4.x.
conda activate takes 1 MINUTE！！！
problem **solved** after I update tensorflow/keras to 2.8

**Comment by HaleTom** [comment]
@ijstokes did you get a ticket number / tracking URL when you raised the issue?

**Comment by HaleTom** [comment]
@mbargull It seems this was closed by an enthusiastic bot.  Can we please re-open?

**Comment by mbargull** [comment]
> It seems this was closed by an enthusiastic bot.

No, it was closed 3 years after I've pointed out in https://github.com/conda/conda/issues/8609#issuecomment-487884014 that this is most likely caused by the `keras` package and upstream `keras`' import time.

> Can we please re-open?

No, unless further information is provided that shows otherwise, I assume this is a package's problem, not one of the package manager.
https://github.com/conda/conda/issues/8609#issuecomment-1786407534 reports improvements with an updated package.
